Arduino IDE Crashed Soon After "ant run"(WIN7)

According to this guide: [https://github.com/arduino/Arduino/wiki/Building-Arduino]
I succeeded in compiling arduino ide source code, but when I try to launch it using "ant run", it immediately crashes.  What can i do to help?


My OS is WIN7.
jdk installation path: D:\Program Files (x86)\Java\jdk1.8.0_202\
JAVA_HOME = D:\Program Files (x86)\Java\jdk1.8.0_202\
CLASSPATH = .%JAVA_HOME%\lib;%JAVA_HOME%\lib\tools.jar


arduino_debug.exe reports following info.

D:\Arduino\build\windows\work>arduino_debug.exe
java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Native
        at com.sun.jna.platform.win32.Shell32.<clinit>(Shell32.java:32)
        at com.sun.jna.platform.win32.Shell32Util.getFolderPath(Shell32Util.java
:44)
        at com.sun.jna.platform.win32.Shell32Util.getFolderPath(Shell32Util.java
:61)
        at cc.arduino.os.windows.Win32KnownFolders.getLocalAppDataFolder(Win32Kn
ownFolders.java:53)
        at processing.app.windows.Platform.recoverSettingsFolderPath(Platform.ja
va:64)
        at processing.app.windows.Platform.init(Platform.java:54)
        at processing.app.Base.<init>(Base.java:208)
        at processing.app.Base.main(Base.java:151)

I have the same issue on windows 10
I also have the weird security permissions after `ant build`.
Moving `jnidispatch-4.2.2-win32-x86.dll` and then copying it back fixes the issue.
Moving it and then moving it back does not fix the issue. This suggests that some of the properties of that file are incorrect, and they get fixed when copying the file.

I did some testing, and it seems to be caused by the cygwin `unzip` executable. Manually unzipping with that executable gives the weird security permissions (and a run failure for `arduino.exe`), manually running the Git `unzip` gives normal security permissions without issues.

I also tried substituting the cygwin unzip with the git `unzip` to see if `ant` would build it correctly but instead `ant` fails with the following error:
```
unzip:
     [echo] Unzipping windows/avr-gcc-5.4.0-atmel3.6.1-arduino2-i686-w64-mingw32.zip into folder windows/work/hardware/tmp/gcc/

BUILD FAILED
path\to\Arduino\build\build.xml:124: The following error occurred while executing this line:
path\to\Arduino\build\build.xml:118: The following error occurred while executing this line:
path\to\Arduino\build\build.xml:1046: The following error occurred while executing this line:
path\to\Arduino\build\build.xml:1179: The following error occurred while executing this line:
path\to\Arduino\build\build.xml:1256: The following error occurred while executing this line:
path\to\Arduino\build\build.xml:908: exec returned: -1073741515
```
The exit code is the windows exit code for a missing `.dll`

I also tried `ant run` from git bash instead of cygwin, and that worked (IDE starts successfully).

--- a/build/build.xml
+++ b/build/build.xml
@@ -512,7 +512,7 @@
   <!-- Unzip AVR tools -->
   <target name="macosx-build-avr-toolchain" unless="light_bundle">
     <antcall target="avr-toolchain-bundle">
-      <param name="unpack_target" value="untar"/>
+      <param name="unpack_target" value="untar-native"/>
       <param name="gcc_archive_file" value="avr-gcc-${AVRGCC-VERSION}-i386-apple-darwin11.tar.bz2"/>
       <param name="gcc_version" value="${AVRGCC-VERSION}"/>
       <param name="avrdude_archive_file" value="avrdude-${AVRDUDE-VERSION}-i386-apple-darwin11.tar.bz2"/>
@@ -725,7 +725,7 @@
     <antcall target="build-arduino-builder" />
 
     <antcall target="avr-toolchain-bundle">
-      <param name="unpack_target" value="untar"/>
+      <param name="unpack_target" value="untar-native"/>
       <param name="gcc_archive_file" value="avr-gcc-${AVRGCC-VERSION}-armhf-pc-linux-gnu.tar.bz2"/>
       <param name="gcc_version" value="${AVRGCC-VERSION}"/>
       <param name="avrdude_archive_file" value="avrdude-${AVRDUDE-VERSION}-armhf-pc-linux-gnu.tar.bz2"/>
@@ -767,7 +767,7 @@
     <antcall target="build-arduino-builder" />
 
     <antcall target="avr-toolchain-bundle">
-      <param name="unpack_target" value="untar"/>
+      <param name="unpack_target" value="untar-native"/>
       <param name="gcc_archive_file" value="avr-gcc-${AVRGCC-VERSION}-aarch64-pc-linux-gnu.tar.bz2"/>
       <param name="gcc_version" value="${AVRGCC-VERSION}"/>
       <param name="avrdude_archive_file" value="avrdude-${AVRDUDE-VERSION}-aarch64-pc-linux-gnu.tar.bz2"/>
@@ -785,7 +785,7 @@
     <antcall target="build-arduino-builder" />
 
     <antcall target="avr-toolchain-bundle">
-      <param name="unpack_target" value="untar"/>
+      <param name="unpack_target" value="untar-native"/>
       <param name="gcc_archive_file" value="avr-gcc-${AVRGCC-VERSION}-i686-pc-linux-gnu.tar.bz2"/>
       <param name="gcc_version" value="${AVRGCC-VERSION}"/>
       <param name="avrdude_archive_file" value="avrdude-${AVRDUDE-VERSION}-i686-pc-linux-gnu.tar.bz2"/>
@@ -803,7 +803,7 @@
     <antcall target="build-arduino-builder" />
 
     <antcall target="avr-toolchain-bundle">
-      <param name="unpack_target" value="untar"/>
+      <param name="unpack_target" value="untar-native"/>
       <param name="gcc_archive_file" value="avr-gcc-${AVRGCC-VERSION}-x86_64-pc-linux-gnu.tar.bz2"/>
       <param name="gcc_version" value="${AVRGCC-VERSION}"/>
       <param name="avrdude_archive_file" value="avrdude-${AVRDUDE-VERSION}-x86_64-pc-linux-gnu.tar.bz2"/>
@@ -852,7 +852,7 @@
   <target name="build-arduino-builder" unless="no_arduino_builder">
     <delete dir="${staging_folder}/arduino-builder-${platform}" includeemptydirs="true"/>
     <mkdir dir="${staging_folder}/arduino-builder-${platform}"/>
-    <antcall target="untar">
+    <antcall target="untar-native">
       <param name="archive_file" value="./arduino-builder-${platform}-${ARDUINO-BUILDER-VERSION}.tar.bz2" />
       <param name="archive_url" value="https://downloads.arduino.cc/tools/arduino-builder-${platform}-${ARDUINO-BUILDER-VERSION}.tar.bz2" />
       <param name="final_folder" value="${staging_folder}/arduino-builder-${platform}/arduino-builder" />
@@ -914,6 +914,14 @@
   <target name="untar-bz2" depends="untar-unzip-checksum" unless="${archive_file}_installed">
     <echo>Untarring ${archive_file} into folder ${dest_folder}</echo>
     <untar src="${archive_file}" dest="${dest_folder}" compression="bzip2"/>
+  </target>
+  <target name="untar-native" depends="untar-unzip-checksum" unless="${archive_file}_installed">
+    <echo>Untarring ${archive_file} into folder ${dest_folder}</echo>
+    <exec executable="tar" failonerror="true">
+      <arg value="xf"/>
+      <arg value="${archive_file}"/>
+      <arg value="--directory=${dest_folder}"/>
+    </exec>
   </target>
 
   <target name="unzip" depends="untar-unzip-checksum" unless="${archive_file}_installed">
@@ -1005,7 +1013,7 @@
   </target>
 
   <target name="download-launch4j-linux">
-    <antcall target="untar">
+    <antcall target="untar-native">
       <param name="archive_file" value="windows/launch4j-3.9-linux.tgz"/>
       <param name="archive_url" value="https://downloads.arduino.cc/tools/launch4j-3.9-linux.tgz"/>
       <param name="final_folder" value="windows/launcher/launch4j"/>

--- a/build/build.xml
+++ b/build/build.xml
@@ -282,7 +282,7 @@
   <!-- copy hardware folder -->
   <target name="assemble-hardware" unless="light_bundle">
     <mkdir dir="${target.path}/hardware/arduino" />
-    <antcall target="untar">
+    <antcall target="untar-bz2">
       <param name="archive_file" value="avr-${AVRCORE-VERSION}.tar.bz2"/>
       <param name="archive_url" value="https://downloads.arduino.cc/cores/avr-${AVRCORE-VERSION}.tar.bz2"/>
       <param name="final_folder" value="${target.path}/hardware/arduino/avr"/>
@@ -528,9 +528,7 @@
     <delete dir="macosx/work/Arduino.app" />
 
     <!-- Unzip unsigned app into working dir -->
-    <exec executable="unzip" dir="macosx/work" failonerror="true">
-      <arg line="../arduino-${version}-${platform}.zip" />
-    </exec>
+    <unzip src="../arduino-${version}-${platform}.zip" dest="macosx/work"/>
 
     <!-- Unlock keychain file -->
     <exec executable="security" dir="macosx/work" failonerror="true">
@@ -895,26 +893,14 @@
   <!-- Ensure that the tool is downloaded and test checksums, if everything's ok unzip it on the tools folder  -->
   <target name="untar" depends="untar-unzip-checksum" unless="${archive_file}_installed">
     <echo>Untarring ${archive_file} into folder ${dest_folder}</echo>
-    <exec executable="tar" failonerror="true">
-      <arg value="xf"/>
-      <arg value="${archive_file}"/>
-      <arg value="--directory=${dest_folder}"/>
-    </exec>
+    <untar src="${archive_file}" dest="${dest_folder}"/>
+  </target>
+  <target name="untar-bz2" depends="untar-unzip-checksum" unless="${archive_file}_installed">
+    <echo>Untarring ${archive_file} into folder ${dest_folder}</echo>
+    <untar src="${archive_file}" dest="${dest_folder}" compression="bzip2"/>
   </target>
 
   <target name="unzip" depends="untar-unzip-checksum" unless="${archive_file}_installed">
-    <echo>Unzipping ${archive_file} into folder ${dest_folder}</echo>
-    <mkdir dir="${dest_folder}" />
-    <exec executable="unzip" failonerror="true">
-      <arg value="-q" />
-      <arg value="-n" />
-      <arg value="-d" />
-      <arg value="${dest_folder}" />
-      <arg value="${archive_file}" />
-    </exec>
-  </target>
-
-  <target name="unzip-with-ant-task" depends="untar-unzip-checksum" unless="${archive_file}_installed">
     <echo>Unzipping ${archive_file} into folder ${dest_folder}</echo>
     <mkdir dir="${dest_folder}" />
     <unzip src="${archive_file}" dest="${dest_folder}"/>
@@ -994,7 +980,7 @@
   </target>
 
   <target name="download-launch4j-windows">
-    <antcall target="unzip-with-ant-task">
+    <antcall target="unzip">
       <param name="archive_file" value="windows/launch4j-3.9-win32.zip"/>
       <param name="archive_url" value="https://downloads.arduino.cc/tools/launch4j-3.9-win32.zip"/>
       <param name="final_folder" value="windows/launcher/launch4j"/>
@@ -1068,7 +1054,7 @@
 
     <delete dir="${staging_folder}/arduino-builder-windows" includeemptydirs="true"/>
     <mkdir dir="${staging_folder}/arduino-builder-windows"/>
-    <antcall target="unzip-with-ant-task">
+    <antcall target="unzip">
       <param name="archive_file" value="./arduino-builder-windows-${ARDUINO-BUILDER-VERSION}.zip" />
       <param name="archive_url" value="https://downloads.arduino.cc/tools/arduino-builder-windows-${ARDUINO-BUILDER-VERSION}.zip" />
       <param name="final_folder" value="${staging_folder}/arduino-builder-windows/arduino-builder.exe" />
@@ -1087,42 +1073,33 @@
     </copy>
     <delete dir="${staging_folder}/arduino-builder-windows" includeemptydirs="true"/>
 
-    <exec executable="unzip" failonerror="true">
-      <arg value="-q" />
-      <arg value="-n" />
-      <arg value="-j" />
-      <arg value="-d" />
-      <arg value="windows/work/lib" />
-      <arg value="../arduino-core/lib/jna-4.2.2.jar" />
-      <arg value="com/sun/jna/win32-x86/jnidispatch.dll" />
-    </exec>
+    <unzip src="../arduino-core/lib/jna-4.2.2.jar" dest="windows/work/lib">
+      <patternset>
+        <include name="com/sun/jna/win32-x86/jnidispatch.dll"/>
+      </patternset>
+      <mapper type="flatten"/>
+    </unzip>
     <move file="windows/work/lib/jnidispatch.dll" tofile="windows/work/lib/jnidispatch-4.2.2-win32-x86.dll" />
     <antcall target="make-file-executable">
       <param name="file" value="windows/work/lib/jnidispatch-4.2.2-win32-x86.dll" />
     </antcall>
 
-    <exec executable="unzip" failonerror="true">
-      <arg value="-q" />
-      <arg value="-n" />
-      <arg value="-j" />
-      <arg value="-d" />
-      <arg value="windows/work/lib" />
-      <arg value="../arduino-core/lib/jssc-2.8.0-arduino3.jar" />
-      <arg value="libs/windows/jSSC-2.8_x86.dll" />
-    </exec>
+    <unzip src="../arduino-core/lib/jssc-2.8.0-arduino3.jar" dest="windows/work/lib">
+      <patternset>
+        <include name="libs/windows/jSSC-2.8_x86.dll"/>
+      </patternset>
+      <mapper type="flatten"/>
+    </unzip>
     <move file="windows/work/lib/jSSC-2.8_x86.dll" tofile="windows/work/lib/jSSC-2.8_x86.dll" />
     <antcall target="make-file-executable">
       <param name="file" value="windows/work/lib/jSSC-2.8_x86.dll" />
     </antcall>
-    <exec executable="unzip" failonerror="true">
-      <arg value="-q" />
-      <arg value="-n" />
-      <arg value="-j" />
-      <arg value="-d" />
-      <arg value="windows/work/lib" />
-      <arg value="../arduino-core/lib/jssc-2.8.0-arduino3.jar" />
-      <arg value="libs/windows/jSSC-2.8_x86_64.dll" />
-    </exec>
+    <unzip src="../arduino-core/lib/jssc-2.8.0-arduino3.jar" dest="windows/work/lib">
+      <patternset>
+        <include name="libs/windows/jSSC-2.8_x86_64.dll"/>
+      </patternset>
+      <mapper type="flatten"/>
+    </unzip>
     <move file="windows/work/lib/jSSC-2.8_x86_64.dll" tofile="windows/work/lib/jSSC-2.8_x86_64.dll" />
     <antcall target="make-file-executable">
       <param name="file" value="windows/work/lib/jSSC-2.8_x86_64.dll" />

--- a/build/build.xml
+++ b/build/build.xml
@@ -114,6 +114,22 @@
     <include name="arduino-core/lib/*.jar" />
   </fileset>
 
+  <!-- Remove all default excludes for directory tasks (see https://ant.apache.org/manual/dirtasks.html#defaultexcludes) -->
+  <defaultexcludes remove="**/.git"/>
+  <defaultexcludes remove="**/.git/**"/>
+  <defaultexcludes remove="**/.gitattributes"/>
+  <defaultexcludes remove="**/.gitignore"/>
+  <defaultexcludes remove="**/.gitmodules"/>
+  <defaultexcludes remove="**/.hg"/>
+  <defaultexcludes remove="**/.hg/**"/>
+  <defaultexcludes remove="**/.hgignore"/>
+  <defaultexcludes remove="**/.hgsub"/>
+  <defaultexcludes remove="**/.hgsubstate"/>
+  <defaultexcludes remove="**/.hgtags"/>
+  <defaultexcludes remove="**/.bzr"/>
+  <defaultexcludes remove="**/.bzr/**"/>
+  <defaultexcludes remove="**/.bzrignore"/>
+
   <target name="build" description="Build Arduino.">
     <antcall target="${platform}-build" />
 

